src/irmin-pack/inode.ml

```diff
@@ -759,68 +796,99 @@ struct
     include T
     module I = Val_impl
 
-    type t = { find : H.t -> I.t option; v : I.t }
-
-    let pred t = I.pred t.v
-    let niet _ = assert false
-
-    let v l =
-      let v = I.v l in
-      { find = niet; v }
-
-    let list ?offset ?length t = I.list ~find:t.find ?offset ?length t.v
-    let empty = { find = niet; v = I.empty }
-    let is_empty t = I.is_empty t.v
-    let find t s = I.find ~find:t.find t.v s
-
-    let add t s v =
-      let v = I.add ~find:t.find t.v s v in
-      if v == t.v then t else { t with v }
-
-    let remove t s =
-      let v = I.remove ~find:t.find t.v s in
-      if v == t.v then t else { t with v }
-
-    let pre_hash_i = Irmin.Type.(unstage (pre_hash I.t))
+    type t =
+      | Total of I.direct_ptr I.layout * I.direct_ptr I.t
```

Comment:
```suggestion
      | Total of I.direct_ptr I.t
```

Reply:
Yes, the layout doesn't have to be stored in `Total` and `Truncated` since it is just a singleton. I just kept if for symmetry with `Partial`, but I can remove that.

Reply:
Didn't realised that you can't remove it in case of `Partial`; symmetry is better indeed.

Reply:
I also think that removing it from `Total` and `Truncated` wouldn't change their size in memory because the size of a variant in memory is the size of the largest tag (i.e. `Partial`). Can someone confirm that?

Reply:
Argumentless variants are always unboxed single words. Variants with _n_ arguments are heap blocks with _n_ fields, so _n + 2_ words in total (pointer + header + _n_ fields).

```ocaml
# type t = Zero | One of int | Two of int * int
  Obj.is_int (Obj.repr Zero) ;;
- : bool = true

# Obj.reachable_words (Obj.repr (One 1)) ;;
- : int = 2 (* header + field *)

# Obj.reachable_words (Obj.repr (Two (1, 2))) ;;
- : int = 3 (* header + 2 fields *)
```

... unless the variant has exactly 1 constructor with exactly 1 argument, in which case it may be unboxed if `-unboxed-types` is set. So: removing from `Total` and `Truncated` _would_ change their in-memory size, and in general two cases of the same variant have no effect on each others' sizes.

Reply:
Thank you :)

Reply:
I chose to remove the unnecessary values, breaking symmetry with Partial